### DISCUSSION TITLE
Crypto Currency Scanners

### DISCUSSION THREAD

**Discussion by bobbyK12345** [simsong/bulk_extractor#474] [Ideas]
Could BE be used to scan an image for a Bitcoin address, Ethereum, etc.? or would a scanner need to be compiled for such a purpose?

**Comment by simsong** [comment]
Sure. How do you want to validate them?
It could be a special-purpose scanner (other people have built them), or done with a regular expression. If it is a special-purpose scanner, it could also run any checksum algorithms.

**Comment by bobbyK12345** [comment]
Good question - I don't know that crypto currency addresses have to meet a checksum (do they?), but if I can run it through a regex search I will try that first.
I was thinking of trying to build a scanner as part of the financial scanner, but without a checksum or anyway to validate that would probably result in too many responses.
Thanks for the response!
Bob

    On Friday, June 28, 2024 at 02:27:23 PM CDT, Simson L. Garfinkel ***@***.***> wrote:  
 
 


Sure. How do you want to validate them?
It could be a special-purpose scanner (other people have built them), or done with a regular expression. If it is a special-purpose scanner, it could also run any checksum algorithms.

**Reply by simsong** [reply]
You will be very unhappy with a regex that searches for hexadecimal numbers in a row unless you have other formatting information that can be keyed off of.

**Comment by bobbyK12345** [comment]
What would you recommend?  how would you search for crypto addresses?
    On Thursday, July 11, 2024 at 02:26:09 PM CDT, Simson L. Garfinkel ***@***.***> wrote:  
 
 


You will be very unhappy with a regex that searches for hexadecimal numbers in a row unless you have other formatting information that can be keyed off of.

**Comment by simsong** [comment]
I would propose a two step process:

1 - Identify potential candidates with a regular expression.
2 - For crypto addresses that have a checksum or some other mathematical check, use it. Otherwise, look for other contextual information, like "bitcoin" or "BTC" or ₿ or the word "address" or "wallet".

**Comment by simsong-codex** [comment]
I would keep this as a **two-stage system**, but redefine the second stage carefully.

An address can be (1) **well formed** (alphabet, length, checksum/version where applicable); (2) **plausible for a specific chain/network**; or (3) **observed on-chain** (or present/funded on an account chain). Only (1), and much of (2), is decidable from the image. An Internet API cannot establish ownership or whether an address can receive funds. A never-used Bitcoin address and an unfunded but valid Solana public key should not be called invalid. Thus the second pass should report *observed / not observed / unknown*, with chain, network, endpoint, timestamp, and error state - not Boolean "valid."

## Proposed design

**Pass 1: offline scanner.** Tokenize candidates with conservative boundaries; recognize an explicit address family; decode it; verify its checksum and permitted version/network; then emit \`currency_family\`, \`network\`, \`encoding\`, \`checksum_status\`, and confidence. This is the normal forensic result and must work without network access.

Avoid a generic hexadecimal regex: 40 hex digits occur in hashes, IDs, and source code. In particular, bare lower-case Ethereum-style addresses have no mandatory checksum. Require \`0x\`, an EIP-55 mixed-case checksum when present, or useful nearby context before assigning high confidence. EIP-55 has only about 15 check bits on average.

**Pass 2: opt-in report enrichment, not a scanner callback.** Deduplicate Pass-1 results and query a locally operated full node/indexer where possible; allow a configured third-party provider only explicitly. Cache, batch, rate-limit, and make endpoints/API keys external configuration. A default call-out would disclose candidate addresses and timing to a third party, impair reproducibility, and make imaging depend on network/rate limits. Lookup failure must remain *unknown*, never invalid. For a token asset, the lookup must specify both host chain and token contract/mint; an address's text alone does not make it a USDT or USDC address.

## Scope: practical top-20 starting set

Market-cap ordering moves and stablecoins span chains, so implement address *families*, not 20 unrelated scanners. Estimates include implementation, vectors, and documentation; add roughly 1-3 days per supported API/network adapter.

| Currency / family | Pass-1 validation | Pass-2 interpretation | Effort |
| --- | --- | --- | --- |
| **Bitcoin (BTC)** | Base58Check legacy/P2SH and Bech32/Bech32m SegWit/Taproot; mainnet versions | address/UTXO/history; zero history is not invalid | Medium, ~3-5 days |
| **Ethereum (ETH)** | \`0x\` + 20 bytes; validate EIP-55 only when mixed case; all-lower/upper permitted | account/code/balance/history; chain ID required | Medium, ~2-3 days offline; ~1-2 per RPC |
| **Tether (USDT), USD Coin (USDC), Chainlink (LINK), Lido stETH** | Mostly token contracts, not independent address encodings | query selected contract/mint and chain; no address-text-only claim | Low scanner increment; Medium API/config |
| **BNB** | BNB Smart Chain uses EVM form; legacy Beacon forms separate | BSC-specific lookup | Low after EVM |
| **XRP (XRP)** | XRPL Base58 alphabet + 4-byte checksum; classic \`r...\` and ideally X-addresses carrying tags | funded-account/ledger state; tag is material | Low-Medium, ~1-2 days |
| **Solana (SOL)** | Base58-decode 32 bytes; optional on-curve test; no BTC-like typo checksum | owner/type/mint state; unfunded on-curve key still valid | Medium, ~2-3 days |
| **TRON (TRX)** | Base58Check + TRON version; hex is contextual | account/contract state | Low-Medium, ~1-2 days |
| **Dogecoin (DOGE), Litecoin (LTC)** | BTC-family Base58Check with coin versions; LTC has several forms | UTXO/history | Low incremental effort after BTC |
| **Bitcoin Cash (BCH)** | CashAddr checksum + optional legacy Base58 | UTXO/history | Low-Medium, ~1-2 days |
| **Cardano (ADA)** | Bech32 plus header/network/payload semantics; payment/stake differ | UTxO/address state | Medium-High, ~3-5 days |
| **Stellar (XLM)** | StrKey/Base32 CRC16; distinguish public \`G...\`, muxed \`M...\`, key types | account state | Low-Medium, ~1-2 days |
| **Avalanche (AVAX)** | C-Chain EVM; X/P Chains separate Bech32-like formats | lookup must name C/X/P | Medium, ~2-3 days |
| **Sui (SUI)** | 32-byte hex identifiers; weak text discrimination | object/account evidence, not ownership | Medium, ~1-2 days + context policy |
| **Hyperliquid (HYPE)** | EVM-style address | Hyperliquid-specific account lookup | Low after EVM; separate adapter |
| **Hedera (HBAR)** | numeric \`shard.realm.num\` and EVM aliases differ; collisions without context | account state | Medium, ~2-3 days |
| **TON (TON)** | friendly/raw address encoding with flags/checksum | account state | Medium, ~2-3 days |
| **Monero (XMR)** | Base58 blocks, network/type bytes, checksum; standard/integrated/subaddresses | public lookup intentionally limited; cannot reveal normal history | Medium-High, ~3-5 days; limited Pass-2 value |

This covers current large-cap currencies while avoiding the error of treating USDT, USDC, LINK, stETH, and BNB as wholly separate patterns. It also keeps syntax confidence distinct from chain observation, which matters in forensic reporting.

Recommended order:

1. BTC, EVM (ETH plus EVM tokens/chains), XRP, SOL.
2. BTC derivatives, TRON, BCH, XLM.
3. ADA, AVAX, TON, HBAR, SUI.
4. XMR only if offline detection value justifies its complexity.

Initially ship Pass 1 for group 1 with positive/negative test vectors and contextual false-positive tests. Add Pass 2 as a separate command or enrichment mode, beginning with a configured Bitcoin node and Ethereum-compatible JSON-RPC endpoint.

References: [EIP-55](https://eips.ethereum.org/EIPS/eip-55), [Solana address verification](https://solana.com/docs/payments/send-payments/verify-address), and [XRPL address formats](https://xrpl.org/docs/references/protocol/data-types/basic-data-types).

**Comment by simsong-codex** [comment]
### Clarifying effort, false positives, lookup cost, and code sharing

In the earlier table, effort is developer effort to get a maintainable, evidence-backed feature, not the CPU time to scan an image. It includes specifying valid mainnet forms; decoding/checksum/network rules; public positive and negative vectors; boundary/context false-positive tests; documentation; and report integration. API-adapter effort separately includes configuration, batching/caching/rate limits, response interpretation, and timeout/error handling.

Thus BTC at 3-5 days means one engineer can normally cover legacy Base58Check plus Bech32/Bech32m and good tests in that time. EVM at 2-3 days means shared work for Ethereum, BSC and other EVM chains, not 2-3 days per ERC-20. Cardano and Monero are higher due to their valid address forms and network/type semantics, not scan speed.

## False-positive expectations

There is no defensible universal percentage until we run representative labelled disk images: source code, hashes, logs, encoded blobs, and wallet applications have very different base rates. We should measure and publish candidates per GiB by confidence tier. The following are conditional random-candidate acceptance estimates, not a rate across all strings in a disk image:

| Family | After appropriate prefix/boundary rule | Likely forensic risk |
| --- | --- | --- |
| BTC Base58Check; XRP; TRON; DOGE/LTC legacy; Monero | typically a 32-bit checksum: about 1 random shape match in 2^32 passes | Very low after decode/checksum; remaining risk is incorrect family/network/context |
| BTC Bech32/Bech32m, BCH CashAddr, ADA Bech32 | about a 30-bit checksum | Very low after full decode/checksum |
| XLM StrKey and common TON friendly form | 16-bit checksum: about 1 in 65,536 shape matches passes | Low, but material at massive candidate volume |
| ETH/EVM mixed-case EIP-55 | published random mistype false-acceptance rate is 0.0247% | Low when the checksum case is present |
| ETH/EVM all-lower/all-upper; Sui hex | no mandatory textual checksum | High for a plain hex rule; require strong boundaries/context and lower confidence |
| Solana Base58 32-byte key | no typo-catching address checksum | Medium-high in Base58-rich data; length/decode alone is insufficient |
| Hedera numeric ID | no checksum and ordinary decimal punctuation | High absent nearby account/HBAR/wallet/schema context |

The scanner should record why it accepted a candidate: checksum, prefix, context, and/or curve check. That permits a user to set report thresholds.

## Lookup cost

The major forensic costs are usually privacy, rate limits, and operations rather than CPU:

* A local node/indexer has no vendor per-query charge and best confidentiality, but requires storage, synchronization, maintenance, and a separate operational setup per chain.
* Hosted RPC/indexers commonly provide a quota-backed free tier but have rate and historical-data limits. As one current example, QuickNode lists 10M API credits/month and 15 requests/second for its free trial; its paid plan starts at $49/month and additional credits at $0.62 per million. Actual cost per address depends on the chain and query: a balance/account read may be one request, but token holdings/history can require several calls or an indexer.
* Deduplicate first, then cache each result using chain, network, address, query type, and block height/time. Batch where supported and set a job lookup budget. Do not spend lookups on weak Pass-1 candidates. A job with 10,000 strong unique addresses is roughly 10,000 simple calls; token/history enrichment can multiply that number.

Store endpoint/local-node identity, query time, chain tip or block height, request count, and result state. A lookup failure is unknown, never invalid.

## Code sharing and parallel execution

Use one crypto scanner with a small common address-validation library and thin family adapters, rather than 20 scanners with copied regexes and network code.

Shared components:

* Candidate tokenizer and context scorer.
* Family registry: prefixes, permitted length, network, and confidence policy.
* Generic codecs: Base58Check, Bech32/Bech32m, CashAddr, Base32/CRC, hex, and EIP-55.
* Normalized finding and a separate lookup queue/cache/rate limiter/provenance recorder.
* Pure local validators that can execute concurrently in bulk_extractor's normal parallel scanning model; remote lookup must not block image scanning.

Examples of reuse:

* BTC, DOGE, LTC, and legacy BCH share Base58Check; BCH adds CashAddr.
* BTC Bech32/Bech32m, Cardano, and Avalanche X/P forms share Bech32 primitives, while each adapter owns its HRP, version, and payload rules.
* ETH, BSC, Avalanche C-Chain, Hyperliquid EVM, and EVM-hosted USDT/USDC/LINK/stETH share 20-byte hex parsing and EIP-55. The adapter supplies chain ID and, for tokens, contract address.
* XRP, TRON, and other systems can share Base58 machinery but must not share alphabet/version assumptions.
* Solana and Monero can share Base58 decoding only; their curve, block, and checksum semantics stay separate.

Run common tokenization once, dispatch each candidate to registered offline validators concurrently, and put Pass 2 in a bounded worker pool or later enrichment command. Before implementation, add a benchmark corpus and publish observed confidence-tier counts per GiB; that is how we derive a useful false-positive rate.

Sources: [EIP-55](https://eips.ethereum.org/EIPS/eip-55) and [QuickNode pricing](https://www.quicknode.com/pricing).

**Comment by simsong-codex** [comment]
### Matrix addendum: addresses, public keys, private keys, and encrypted wallets

Every currency in the earlier matrix is based on public-key cryptography and therefore has private keys. That does **not** mean a scanner can find or validate a private key from an address: an address is normally a hash, encoding, or chain-specific identifier derived from a public key. The scanner must report these separately.

��y��y�Password brute force requiref��y��y� below means only: an authorized examiner has an **encrypted wallet/key container** and cannot otherwise decrypt it. It is not an operation on an address, public key, chain API, or ordinary disk string. A modern private key cannot be feasibly derived from its address/public key by brute force.

| Currency / family | Address / account identifier to scan | Raw public-key form | Private/secret material to recognize separately | Password cracking required? |
| --- | --- | --- | --- | --- |
| BTC; DOGE; LTC; BCH | Base58Check, Bech32/Bech32m, CashAddr as applicable | compressed/uncompressed secp256k1 public key, normally not displayed as an address | 32-byte secp256k1 key; WIF; BIP-32 extended private keys; BIP-39 seed phrase | No for an address/WIF; only for an encrypted wallet/keystore/backup |
| ETH; BSC; Avalanche C; Hyperliquid EVM; EVM tokens including USDT/USDC/LINK/stETH | 20-byte EVM address (0x + 40 hex) | secp256k1 public key, usually not retained/displayed as the address | 32-byte secp256k1 key; JSON keystore; BIP-32/BIP-39 material | Only for encrypted JSON keystore/wallet/backup |
| XRP | classic r-address or X-address | secp256k1 or Ed25519 public key | family-specific private key/seed encoding | Only for encrypted wallet/seed backup |
| Solana | 32-byte Base58 account/public-key address | the address is normally the Ed25519 public key | 32-byte seed or secret-key material; commonly 64-byte keypair export; BIP-39 seed phrase | Only for encrypted wallet/seed backup; a bare 32-byte Base58 string is potentially ambiguous |
| TRON | Base58Check account address; sometimes hex presentation | secp256k1 public key | 32-byte secp256k1 key; mnemonic/keystore | Only for encrypted wallet/backup |
| Cardano | Bech32 payment/stake address | Ed25519 public key / extended public key | Ed25519 private or extended private key; recovery phrase | Only for encrypted wallet/backup |
| Stellar | StrKey G public address or M muxed address | Ed25519 public key encoded as address | StrKey secret seed beginning S | No for a bare S secret; only if its container is encrypted |
| Avalanche X/P | chain-specific Bech32 account form | secp256k1 public key | 32-byte secp256k1 key; mnemonic/keystore | Only for encrypted wallet/backup |
| Sui | 32-byte hex address | Ed25519/secp256k1/etc. public key depending on scheme | scheme-tagged private key/seed/mnemonic | Only for encrypted wallet/backup |
| Hedera | numeric shard.realm.num ID or EVM alias | ED25519 or ECDSA public key | corresponding private key, seed, or keystore | Only for encrypted wallet/backup |
| TON | friendly/raw account address | Ed25519 public key (account address is contract/account derived) | Ed25519 private key/seed/mnemonic | Only for encrypted wallet/backup |
| Monero | standard, integrated, or subaddress | spend and view public keys; address encodes public components | private spend key and private view key; mnemonic seed | Only for encrypted wallet/keys; do not query a secret-bearing finding |

Implementation policy:

1. Classify each finding as `address`, `public_key`, `private_key`, `seed_phrase`, `encrypted_wallet`, or `ambiguous`.
2. Only high-confidence `address` findings may enter Pass 2. Never transmit public keys, private keys, seed phrases, encrypted wallets, or ambiguous candidates to a remote service.
3. Treat secret-bearing results as high-sensitivity evidence: restrict/redact normal report output and avoid logging their full value.
4. A 32-byte hex/Base58 value is often ambiguous between a key, hash, seed, ID, or public key. Do not label it a private key solely by length; require a recognized serialization, a wallet/keystore structure, or strong surrounding context.
5. For encrypted wallet containers, report the encryption/container type and"��y��y�encrypted secret material presen{�u����]; do not make automated recovery or password-guessing a normal scanner feature. Any authorized recovery workflow is separate, audited, rate/budget-limited, and governed by the applicable evidence-handling policy.

This makes the key distinction explicit: offline address validation checks a public identifier; encrypted-wallet recovery is a separate evidence workflow; and no feasible brute-force operation turns an address into its private key.
